### PR TITLE
Include local-functions in the go-to index.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/39083

These have risen in priority, esp. as they may be used a lot in a top-level program.

### CODE REVIEW

**Comment by ryzngard** [comment]
Perf concerns aside, I have some usability concerns here. There's currently no way to filter out local functions from other members. That could easily mean that common local function names that coincide with member names will be hard to sort through...

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide this is ready for review.

File: src/Workspaces/VisualBasic/Portable/FindSymbols/VisualBasicDeclaredSymbolInfoFactoryService.vb [hunk 1]
```diff
@@ -126,34 +127,79 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.FindSymbols
...
         End Function
 
+        Protected Overrides Sub AddLocalFunctionInfos(compilationUnit As CompilationUnitSyntax, stringTable As StringTable, declaredSymbolInfos As ArrayBuilder(Of DeclaredSymbolInfo), cancellationToken As CancellationToken)
+            ' VB doesn't have local functions.
+        End Sub
+
+        Protected Overrides Sub AddLocalFunctionInfos(node As StatementSyntax, stringTable As StringTable, declaredSymbolInfos As ArrayBuilder(Of DeclaredSymbolInfo), containerDisplayName As String, fullyQualifiedContainerName As String, cancellationToken As CancellationToken)
+            ' VB doesn't have local functions.
+        End Sub
         Protected Overrides Sub AddSynthesizedDeclaredSymbolInfos(container As SyntaxNode, memberDeclaration As StatementSyntax, stringTable As StringTable, declaredSymbolInfos As ArrayBuilder(Of DeclaredSymbolInfo), containerDisplayName As String, fullyQualifiedContainerName As String, cancellationToken As CancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/FindSymbols/VisualBasicDeclaredSymbolInfoFactoryService.vb:137]:
```suggestion

        Protected Overrides Sub AddSynthesizedDeclaredSymbolInfos(container As SyntaxNode, memberDeclaration As StatementSyntax, stringTable As StringTable, declaredSymbolInfos As ArrayBuilder(Of DeclaredSymbolInfo), containerDisplayName As String, fullyQualifiedContainerName As String, cancellationToken As CancellationToken)
```

File: src/Workspaces/Core/Portable/LanguageServices/DeclaredSymbolFactoryService/AbstractDeclaredSymbolInfoFactoryService.cs [hunk 2]
```diff
@@ -62,8 +64,14 @@ protected AbstractDeclaredSymbolInfoFactoryService()
...
+        protected abstract DeclaredSymbolInfo? GetTypeDeclarationInfo(
+            SyntaxNode container, TTypeDeclarationSyntax typeDeclaration, StringTable stringTable, string containerDisplayName, string fullyQualifiedContainerName);
+        protected abstract DeclaredSymbolInfo GetEnumDeclarationInfo(
+            SyntaxNode container, TEnumDeclarationSyntax enumDeclaration, StringTable stringTable, string containerDisplayName, string fullyQualifiedContainerName);
+        protected abstract void AddMemberDeclarationInfos(
+            SyntaxNode container, TMemberDeclarationSyntax memberDeclaration, StringTable stringTable, ArrayBuilder<DeclaredSymbolInfo> declaredSymbolInfos, string containerDisplayName, string fullyQualifiedContainerName);
+        protected abstract void AddLocalFunctionInfos(
+            TMemberDeclarationSyntax memberDeclaration, StringTable stringTable, ArrayBuilder<DeclaredSymbolInfo> declaredSymbolInfos, string containerDisplayName, string fullyQualifiedContainerName, CancellationToken cancellationToken);
```
Comment (akhera99) [inline src/Workspaces/Core/Portable/LanguageServices/DeclaredSymbolFactoryService/AbstractDeclaredSymbolInfoFactoryService.cs:74]: I think this makes a lot more sense to read than what it was previously

**Review by akhera99 — APPROVED** [review]
Looks alright to me

**Pull request merged**